internal/remediation/override.go

```diff
@@ -279,6 +279,59 @@ func getVersionsGreater(ctx context.Context, cl client.DependencyClient, vk reso
 	return versions[offset:], nil
 }
 
+// comparisonFunctionWithWorkarounds returns a version comparison function with special behaviour for specific packages,
+// producing more desirable ordering using non-standard comparison.
+// TODO: Move this and make it re-usable for other remediation strategies & osv-scanner update.
+func comparisonFunctionWithWorkarounds(vk resolve.VersionKey) func(resolve.Version, resolve.Version) int {
+	sys := vk.Semver()
+
+	if vk.System == resolve.Maven && vk.Name == "com.google.guava:guava" {
+		// com.google.guava:guava has 'flavors' with versions ending with -jre or -android.
+		// https://github.com/google/guava/wiki/ReleasePolicy#flavors
+		// To preserve the flavor in updates, we make the opposite flavor considered the earliest versions.
+
+		// Old versions have '22.0' and '22.0-android', and even older version don't have any flavors.
+		// Only check for the android flavor, and assume its jre otherwise.
+		wantAndroid := strings.HasSuffix(vk.Version, "-android")
+		return func(a, b resolve.Version) int {
+			aIsAndroid := strings.HasSuffix(a.Version, "-android")
+			bIsAndroid := strings.HasSuffix(b.Version, "-android")
+
+			if aIsAndroid == bIsAndroid {
+				return sys.Compare(a.Version, b.Version)
+			}
+
+			if aIsAndroid == wantAndroid {
+				return 1
+			}
+
+			return -1
+		}
+	}
+
+	if vk.System == resolve.Maven && strings.HasPrefix(vk.Name, "commons-") {
```

Comment:
Are these versions in maven-metadata.xml as well ? 

e.g. I see the problematic  [20030203.000550](https://deps.dev/maven/commons-io%3Acommons-io/20030203.000550) for `commons-io:commons-io`, but it doesn't exist in [maven-metadata.xml](https://repo.maven.apache.org/maven2/commons-io/commons-io/maven-metadata.xml)

Is checking maven-metadata.xml the proper fix for this here? If so, let's add a TODO.

Reply:
> Are these versions in maven-metadata.xml as well ?

In nearly all cases, no. I did find [`commons-primitives:commons-primitives`](https://repo1.maven.org/maven2/commons-primitives/commons-primitives/maven-metadata.xml) and [`commons-sql:commons-sql`](https://repo1.maven.org/maven2/commons-sql/commons-sql/maven-metadata.xml) with both formats, and [`commons-compress:commons-compress`](https://repo1.maven.org/maven2/commons-compress/commons-compress/maven-metadata.xml) has only date-based versions. So it's not guaranteed.

> Is checking maven-metadata.xml the proper fix for this here?

Checking the metadata file (which is a necessity for the native registry client) will usually exclude those versions from the list, but we don't do that for the deps.dev API.
If a `pom.xml` was already using e.g. `commons-io:commons-io` version `20030203.000550`, we still would probably want to upgrade it through `0.1` -> `1.4` -> `2.16.1`.

That said, I don't think I've seen any OSV/GHSA records that correctly list those versions as vulnerable :shrug:

Reply:
Ack, so we can't reliably rely on this in any case. Also great point re still wanting to support the proper upgrade path.